packages/verified-fetch/src/index.ts

```diff
@@ -0,0 +1,339 @@
+/**
+ * @packageDocumentation
+ *
+ * `@helia/verified-fetch` is a library that provides a fetch-like API for fetching trustless content from IPFS and
+ * verifying it.
+ *
+ * This library should act as a replacement for the `fetch()` API for fetching content from IPFS, and will return a
+ * [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) object that can be used in a similar manner
+ * to the `fetch()` API. This means browser and HTTP caching inside browser main threads, web-workers, and service
+ * workers, as well as other features of the `fetch()` API should work in a way familiar to developers.
+ *
+ * Exports a `createVerifiedFetch` function that returns a `fetch()` like API method {@link Helia} for fetching IPFS content.
+ *
+ * You may use any supported resource argument to fetch content:
+ *
+ * - CID instance
+ * - IPFS URL
+ * - IPNS URL
+ *
+ * @example
+ *
+ * ```ts
+ * import { createVerifiedFetch } from '@helia/verified-fetch'
+ *
+ * const fetch = await createVerifiedFetch({
+ *  gateways: ['mygateway.info', 'trustless-gateway.link']
+ *})
+ *
+ * const resp = await fetch('ipfs://bafy...')
+ *
+ * const json = await resp.json()
+ *```
+ *
+ *
+ * @example Using a CID instance to fetch JSON
+ *
+ * ```typescript
+ * import { createVerifiedFetch } from '@helia/verified-fetch'
+ * import { CID } from 'multiformats/cid'
+ *
+ * const fetch = await createVerifiedFetch({
+ *  gateways: ['http://mygateway.info', 'http://trustless-gateway.link']
+ * })
+ *
+ * const cid = CID.parse('bafyFoo') // some image file
+ * const response = await fetch(cid)
+ * const json = await response.json()
+ * ```
+ *
+ * @example Using ipfs protocol to fetch an image
+ *
+ * ```typescript
+ * import { createVerifiedFetch } from '@helia/verified-fetch'
+ *
+ * const fetch = await createVerifiedFetch({
+ *  gateways: ['http://mygateway.info', 'http://trustless-gateway.link']
+ * })
+ * const response = await fetch('ipfs://bafyFoo') // CID for some image file
+ * const blob = await response.blob()
+ * ```
+ *
+ * @example Using ipns protocol to fetch a video
+ *
+ * ```typescript
+ * import { createVerifiedFetch } from '@helia/verified-fetch'
+ *
+ * const fetch = await createVerifiedFetch({
+ *  gateways: ['http://mygateway.info', 'http://trustless-gateway.link']
+ * })
+ * const response = await fetch('ipns://mydomain.com/path/to/video.mp4')
```

Comment:
I think we should avoid video here, because videos require efficient seeking, and that can be done with range requests, which we don't have yet.

Let's make this about streaming generic "big file" instead:

```suggestion
 * @example Using ipns protocol to stream a big file
 *
 * ```typescript
 * import { createVerifiedFetch } from '@helia/verified-fetch'
 *
 * const fetch = await createVerifiedFetch({
 *  gateways: ['http://mygateway.info', 'http://trustless-gateway.link']
 * })
 * const response = await fetch('ipns://mydomain.com/path/to/very-long-file.log')
```